cunumeric/_ufunc/ufunc.py

```diff
@@ -154,52 +165,51 @@
 all_dtypes = ["?"] + all_but_boolean
 
 
-def predicate_types_of(dtypes):
+def predicate_types_of(dtypes: Sequence[str]) -> list[str]:
     return [ty + "?" for ty in dtypes]
 
 
-def relation_types_of(dtypes):
+def relation_types_of(dtypes: Sequence[str]) -> list[str]:
     return [ty * 2 + "?" for ty in dtypes]
 
 
-def to_dtypes(chars):
+def to_dtypes(chars: str) -> tuple[np.dtype[Any], ...]:
     return tuple(np.dtype(char) for char in chars)
 
 
 class ufunc:
-    def __init__(self, name, doc, types):
+    def __init__(self, name: str, doc: str, types: dict[Any, str]) -> None:
```

Comment:
I feel we should remove `types` from this parent class, as `types` in each sub-class has a different type (`dict[str, str]` for unary_ufunc, `dict[str, tuple[str, str]]` for multiout_unary_ufunc, and `dict[tuple[str,str],str]` for binary_ufunc)

Reply:
https://github.com/nv-legate/cunumeric/pull/387/commits/f70ae583d855ad9d2f9707e7b8fd4c44be9bf6d2